File: natural_language_processing_saurabh.R
# Natural Language Processing

# Importing the dataset
dataset_original = read.delim('Restaurant_Reviews.tsv', quote = '', stringsAsFactors = FALSE)
#read.delim is used to read tsv file
#quote = '' to disable/ignore quotes
#stringsAsFactors = FALSE means reviews shouldn't be identified as factor/categorical. It must be numeric.

# Cleaning the texts
#install.packages('tm')         #tm package is used for NLP
#install.packages('SnowballC')
library(tm)
library(SnowballC)
corpus = VCorpus(VectorSource(dataset_original$Review))  #VectorSource has parameter which we want to clean ie. the review column
corpus = tm_map(corpus, content_transformer(tolower))   #convert everything to lower case alphabet
corpus = tm_map(corpus, removeNumbers)
corpus = tm_map(corpus, removePunctuation)
corpus = tm_map(corpus, removeWords, stopwords())      #remove stopwords like "the", "a", "an", "in" etc.
corpus = tm_map(corpus, stemDocument)                  #stem is used to find root of the word
corpus = tm_map(corpus, stripWhitespace)

# Creating the Bag of Words model
dtm = DocumentTermMatrix(corpus)     #Creates Sparse Matrix  
dtm = removeSparseTerms(dtm, 0.999)  #only 99.9% most frequently used words are chosen
dataset = as.data.frame(as.matrix(dtm))   #converting independent variables into dataframe
dataset$Liked = dataset_original$Liked    #dependent variable ie. reviews are positive or negative

# Encoding the target feature as factor
dataset$Liked = factor(dataset$Liked, levels = c(0, 1))

# Splitting the dataset into the Training set and Test set
# install.packages('caTools')
library(caTools)
set.seed(123)
split = sample.split(dataset$Liked, SplitRatio = 0.8)
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

# Fitting Random Forest Classification to the Training set
# install.packages('randomForest')
library(randomForest)
classifier = randomForest(x = training_set[-692],
                          y = training_set$Liked,
                          ntree = 10)
#692 is column no. of dependent variable

# Predicting the Test set results
y_pred = predict(classifier, newdata = test_set[-692])

# Making the Confusion Matrix
cm = table(test_set[, 692], y_pred)

File: natural_language_processing_saurabh.py
# Natural Language Processing

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Restaurant_Reviews.tsv', delimiter = '\t', quoting = 3)
#delimiter = '\t' means two columns are seperated by a Tab ie. its a tsv file
#quoting = 3 is the code to ignore double quotes

# Cleaning the texts
import re        #Regular Expression
import nltk      #Natual language Tool Kit
#nltk.download()     #To download all packages of nltk
nltk.download('stopwords')      #use to remove stopwords like “the”, “a”, “an”, “in” etc.
from nltk.corpus import stopwords
from nltk.stem.porter import PorterStemmer    #stem is used to find root of the word 
corpus = []        #Corpus is collection of texts of any type
for i in range(0, 1000):      #no. of reviews is 1000
    review = re.sub('[^a-zA-Z]', ' ', dataset['Review'][i])     #Returns only a-z and A-Z, other non-alphabet letters will be removed and replaced by a white space. 
    review = review.lower()        #convert everything to lower case alphabet
    review = review.split()        #Split the review string into individual words ie. converted to a list.
    ps = PorterStemmer()           #object of PorterStemmer class
    review = [ps.stem(word) for word in review if not word in set(stopwords.words('english'))]
    review = ' '.join(review)
    corpus.append(review)

# Creating the Bag of Words model through Tokenization
from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer(max_features = 1500)   #only 1500 most frequently used words are chosen
X = cv.fit_transform(corpus).toarray()     #Creating Sparse Matrix
Y = dataset.iloc[:, 1].values             #dependent variable ie. reviews are positive or negative

# Splitting the dataset into the Training set and Test set
from sklearn.cross_validation import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.20, random_state = 0)

# Fitting Naive Bayes to the Training set
from sklearn.naive_bayes import GaussianNB
classifier = GaussianNB()
classifier.fit(X_train, Y_train)

# Predicting the Test set results
y_pred = classifier.predict(X_test)

# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(Y_test, y_pred)